File: pdk-ts/src/commands/call.ts
/**
 * `pdk-ts call <pallet> <call> [args...]` — generic extrinsic composer.
 *
 * Every other signing command (`send`, `assets`) hardcodes one specific
 * pallet call. This one doesn't: it looks up ANY pallet.call pair in
 * live metadata via `api.tx[pallet][call].meta.args` — the same kind of
 * runtime introspection `pallets.ts`/`storage.ts`/`explain.ts` already
 * do against `api.runtimeMetadata`/`api.registry.lookup` — validates and
 * coerces the CLI args against that call's real argument types, builds
 * `api.tx[pallet][call](...args)`, signs with the same `--from` URI
 * pattern as every other signing command, and submits through send.ts's
 * `submitExtrinsic`. That's the whole point: it inherits the project's
 * hard invariant (success confirmed ONLY by a positive
 * `system.ExtrinsicSuccess` event, never by the absence of a
 * dispatchError) for free, instead of a new success-detection path that
 * could get the false-success bug wrong again.
 *
 * Discovery is built in, not an afterthought — a generic composer is
 * useless if you have to already know the signature to try it:
 *   `pdk-ts call <pallet>`         → list the pallet's calls + arg types
 *   `pdk-ts call <pallet> <call>`  → show that call's expected args
 *
 * Coercion deliberately covers a narrow, common slice of primitive
 * shapes — AccountId-like (SS58/`//URI`), unsigned integers including
 * `Compact<...>` wrapping (covers Balance/AssetId/etc, which are type
 * aliases over u32/u64/u128 with no distinct runtime shape), bool, and
 * `Bytes`/`Vec<u8>` as 0x-hex. Anything else (structs, enums, Vec<T> of
 * non-bytes, Option<T>, tuples, signed integers) is a hard, named error
 * — never a silent mis-encode.
 */

import pc from 'picocolors';
import {Keyring} from '@polkadot/api';
import type {ApiPromise} from '@polkadot/api';
import {getApi, closeApi} from '../core/chain.js';
import {resolveNode} from '../core/config.js';
import {humanizeChainError} from '../core/errors.js';
import {normaliseUri} from './keys.js';
import {resolveRecipient, submitExtrinsic, type TransferOutcome} from './send.js';

const POT_DECIMALS = 14; // matches send.ts/simulate.ts/accounts.ts — fees are always paid in the native token regardless of which pallet/call is invoked.

export interface CallOptions {
  from?: string;
  node?: string;
  timeout?: string;
  json?: boolean;
  dryRun?: boolean;
}

export interface CallArgMeta {
  name: string;
  type: string;
}

// ---------------------------------------------------------------------
// Argument type classification
// ---------------------------------------------------------------------

export type ArgKind = 'accountId' | 'integer' | 'bool' | 'bytes' | 'unsupported';

/** Peel N layers of `Compact<...>` down to the inner type name. */
function unwrapCompact(rawType: string): string {
  let t = rawType.trim();
  for (;;) {
    const m = /^Compact<(.+)>$/.exec(t);
    if (!m) return t;
    t = m[1].trim();
  }
}

const ACCOUNT_TYPES = /^(AccountId32|AccountId|MultiAddress|Address|LookupSource)$/i;
// Raw primitives, plus the handful of near-universal Substrate numeric
// *aliases* verified live against Portaldot-1002 metadata — `Balance` is
// always a u128 alias, and pallet-assets always aliases its id/amount as
// `AssetId`/`AssetBalance`/`TAssetBalance`. Deliberately NOT a generic
// "ends with Balance/Id" heuristic: e.g. AccountId also ends with "Id" but
// is a 32-byte hash, not a number — matching that broadly would silently
// mis-encode. Anything not on this exact list falls through to 'unsupported'.
const INT_TYPES = /^(u8|u16|u32|u64|u128|u256|Balance|AssetId|AssetBalance|TAssetBalance)$/i;
const BYTES_TYPES = /^(Bytes|Vec<u8>)$/i;

/** Classify a metadata-reported arg type string into a shape we know how to coerce. */
export function classifyType(rawType: string): ArgKind {
  const t = unwrapCompact(rawType);
  if (ACCOUNT_TYPES.test(t)) return 'accountId';
  if (INT_TYPES.test(t)) return 'integer';
  if (/^bool$/i.test(t)) return 'bool';
  if (BYTES_TYPES.test(t)) return 'bytes';
  return 'unsupported';
}

/**
 * Coerce one raw CLI string into what @polkadot/api's tx builder expects
 * for `kind`. Throws a clear, arg-labelled error on bad input rather
 * than letting a malformed value reach SCALE encoding and either fail
 * opaquely or — worse — encode something the user didn't mean.
 */
export function coerceArg(kind: ArgKind, raw: string, label: string, keyring: Keyring): string | boolean {
  switch (kind) {
    case 'accountId':
      // Same SS58-or-`//URI` rule as every other signing command's
      // recipient argument (send.ts's resolveRecipient) — a bare typo'd
      // word is refused rather than silently derived into a black hole.
      return resolveRecipient(keyring, raw);
    case 'integer':
      // Plain non-negative integer only — same strict guard as assets.ts's
      // parseAssetAmount. BigInt() alone accepts "" (→0n) and "0x10" hex,
      // which would silently corrupt whatever the user typed.
      if (!/^\d+$/.test(raw.trim())) throw new Error(`${label}: expected a non-negative integer, got "${raw}"`);
      return BigInt(raw.trim()).toString();
    case 'bool':
      if (!/^(true|false)$/i.test(raw.trim())) throw new Error(`${label}: expected "true" or "false", got "${raw}"`);
      return /^true$/i.test(raw.trim());
    case 'bytes':
      if (!/^0x[0-9a-fA-F]*$/.test(raw.trim())) throw new Error(`${label}: expected 0x-prefixed hex, got "${raw}"`);
      return raw.trim();
    case 'unsupported':
      throw new Error(`${label}: unsupported argument type — pdk-ts call doesn't know how to encode this yet`);
  }
}

// ---------------------------------------------------------------------
// Metadata lookups
// ---------------------------------------------------------------------

// Case-insensitive key match — mirrors storage.ts's findKey so
// `pdk-ts call balances transferKeepAlive` and `Balances transferKeepAlive`
// both resolve the same way storage/pallets commands already do.
function findKey(obj: Record<string, unknown> | undefined, name: string): string | undefined {
  if (!obj) return undefined;
  const target = name.toLowerCase();
  return Object.keys(obj).find((k) => k.toLowerCase() === target);
}

interface BuiltExtrinsic {
  // biome-ignore lint/suspicious/noExplicitAny: sender is a KeyringPair; paymentInfo's return shape varies by chain, we only touch partialFee.
  paymentInfo: (sender: any) => Promise<{partialFee: {toString(): string}}>;
}

interface TxFn {
  meta: {args: Array<{name: {toString(): string}; type: {toString(): string}}>};
  (...args: unknown[]): BuiltExtrinsic;
}

function txPallets(api: ApiPromise): Record<string, Record<string, TxFn>> {
  // api.tx's shape is generated per-chain at connect time; pallet/call
  // names are resolved dynamically at runtime, same as pallets.ts/storage.ts.
  // biome-ignore lint/suspicious/noExplicitAny: see above.
  return api.tx as any;
}

function argMeta(fn: TxFn): CallArgMeta[] {
  return fn.meta.args.map((a) => ({name: a.name.toString(), type: a.type.toString()}));
}

// ---------------------------------------------------------------------
// Command
// ---------------------------------------------------------------------

export async function run(pallet: string, call: string | undefined, args: string[], opts: CallOptions): Promise<void> {
  const node = resolveNode(opts.node);
  try {
    const api = await getApi(node, opts.timeout ? Math.round(Number(opts.timeout) * 1000) : undefined);
    const tx = txPallets(api);
    const palletKey = findKey(tx, pallet);
    if (!palletKey) {
      const msg = `pallet "${pallet}" not found — try \`pdk-ts pallets\` to list`;
      if (opts.json) console.log(JSON.stringify({error: msg}));
      else console.error(pc.red(`\n  ✗ ${msg}\n`));
      await closeApi();
      process.exit(1);
    }

    // Discovery: no call given — list this pallet's calls + arg signatures.
    if (!call) {
      const calls = Object.keys(tx[palletKey])
        .sort()
        .map((c) => ({name: c, args: argMeta(tx[palletKey][c])}));
      if (opts.json) console.log(JSON.stringify({pallet: palletKey, calls}, null, 2));
      else renderCallList(palletKey, calls);
      await closeApi();
      return;
    }

    const callKey = findKey(tx[palletKey], call);
    if (!callKey) {
      const msg = `call "${call}" not found on pallet "${palletKey}" — run \`pdk-ts call ${palletKey}\` to list its calls`;
      if (opts.json) console.log(JSON.stringify({error: msg}));
      else console.error(pc.red(`\n  ✗ ${msg}\n`));
      await closeApi();
      process.exit(1);
    }

    const fn = tx[palletKey][callKey];
    const meta = argMeta(fn);

    // Discovery: call given, no args, and the call actually needs some —
    // show the signature instead of failing blind on an arg-count error.
    if (args.length === 0 && meta.length > 0) {
      if (opts.json) console.log(JSON.stringify({pallet: palletKey, call: callKey, args: meta}, null, 2));
      else renderCallSignature(palletKey, callKey, meta);
      await closeApi();
      return;
    }

    if (args.length !== meta.length) {
      const sig = meta.length ? ` — ${meta.map((m) => `${m.name}: ${m.type}`).join(', ')}` : '';
      throw new Error(`${palletKey}.${callKey} expects ${meta.length} argument(s)${sig}, got ${args.length}`);
    }

    const keyring = new Keyring({type: 'sr25519', ss58Format: 42});
    const coerced = meta.map((m, i) => coerceArg(classifyType(m.type), args[i], `${m.name} (${m.type})`, keyring));

    const extrinsic = fn(...coerced);
    const sender = keyring.addFromUri(normaliseUri(opts.from ?? '//Alice'));

    if (opts.dryRun) {
      await dryRun(api, extrinsic, sender, palletKey, callKey, meta, coerced, Boolean(opts.json));
      await closeApi();
      return;
    }

    const outcome = await submitExtrinsic(api, extrinsic, sender);
    const ok = emitOutcome(palletKey, callKey, outcome, Boolean(opts.json));
    await closeApi();
    if (!ok) process.exit(1);
  } catch (err) {
    const msg = humanizeChainError(err, node);
    if (opts.json) console.log(JSON.stringify({error: msg, endpoint: node}));
    else console.error(pc.red(`\n  ✗ call failed — ${msg}\n`));
    await closeApi();
    process.exit(1);
  }
}

async function dryRun(
  api: ApiPromise,
  extrinsic: BuiltExtrinsic,
  // biome-ignore lint/suspicious/noExplicitAny: KeyringPair.
  sender: any,
  pallet: string,
  call: string,
  meta: CallArgMeta[],
  coerced: Array<string | boolean>,
  json: boolean,
): Promise<void> {
  const info = await extrinsic.paymentInfo(sender);
  const feeRaw = BigInt(info.partialFee.toString());
  const fee = Number(feeRaw) / 10 ** POT_DECIMALS;
  // biome-ignore lint/suspicious/noExplicitAny: system.account is a well-known storage item present on every Substrate chain; shape is stable.
  const accountInfo = await (api.query.system as any).account(sender.address);
  const balanceRaw = BigInt(accountInfo.data.free.toString());
  const balance = Number(balanceRaw) / 10 ** POT_DECIMALS;
  const feasible = balance >= fee;
  const argsOut = meta.map((m, i) => ({name: m.name, type: m.type, value: coerced[i]}));

  if (json) {
    console.log(
      JSON.stringify({dryRun: true, pallet, call, args: argsOut, fee, senderBalance: balance, feasible}),
    );
  } else {
    console.log(`\n  ${pc.bold(`pdk-ts call ${pallet} ${call} --dry-run`)}  ${pc.dim('(not submitted)')}`);
    for (const a of argsOut) console.log(`  ${pc.dim(a.name.padEnd(20))}${String(a.value)}`);
    console.log(`  ${pc.dim('Estimated fee'.padEnd(20))}${fee.toFixed(6)} POT`);
    console.log(`  ${pc.dim('Sender balance'.padEnd(20))}${balance.toLocaleString('en-US')} POT`);
    console.log(
      `  ${pc.dim('Prediction'.padEnd(20))}${feasible ? pc.green('fee payable') : pc.red('would FAIL — insufficient balance for fee')}`,
    );
    console.log(pc.dim('  (fee-payability only — pdk-ts call cannot predict arbitrary pallet dispatch logic)\n'));
  }
  if (!feasible) process.exitCode = 1;
}

function emitOutcome(pallet: string, call: string, outcome: TransferOutcome, json: boolean): boolean {
  const ok = outcome.status === 'success';
  const action = `${pallet}.${call}`;
  if (json) {
    console.log(JSON.stringify({success: ok, status: outcome.status, tx: outcome.txHash, block: outcome.block, error: outcome.error, pallet, call}));
    return ok;
  }
  if (outcome.status === 'success') {
    console.log(pc.green(`\n  ✓ ${action}`));
    console.log(pc.dim(`  tx: ${outcome.txHash}\n`));
  } else if (outcome.status === 'unconfirmed') {
    console.error(pc.yellow(`\n  ⚠ submitted but the outcome couldn't be confirmed — @polkadot/api can't decode Portaldot's result events.`));
    console.error(pc.dim(outcome.txHash ? `  Confirm with Python: pdk debug ${outcome.txHash}\n` : '  No block inclusion within the timeout — check node connectivity and retry.\n'));
  } else {
    console.error(pc.red(`\n  ✗ ${action} failed${outcome.error ? ` — ${outcome.error}` : ''}`));
    console.error(pc.dim(`  tx: ${outcome.txHash}\n`));
  }
  return ok;
}

// ---------------------------------------------------------------------
// Rendering
// ---------------------------------------------------------------------

function renderCallList(pallet: string, calls: Array<{name: string; args: CallArgMeta[]}>): void {
  console.log();
  console.log(pc.bold(`  ${pallet}`) + pc.dim(`  (${calls.length} calls)`));
  console.log();
  for (const c of calls) {
    const sig = c.args.map((a) => `${a.name}: ${a.type}`).join(', ');
    console.log(`  ${pc.green(c.name)}${pc.dim(`(${sig})`)}`);
  }
  console.log();
  console.log(pc.dim(`  pdk-ts call ${pallet} <call>  — show one call's argument types`));
  console.log();
}

function renderCallSignature(pallet: string, call: string, args: CallArgMeta[]): void {
  console.log();
  console.log(pc.bold(`  ${pallet}.${call}`));
  console.log();
  if (args.length === 0) {
    console.log(pc.dim('  (no arguments)'));
  } else {
    const nameW = Math.max(...args.map((a) => a.name.length));
    for (const a of args) {
      const kind = classifyType(a.type);
      const support = kind === 'unsupported' ? pc.yellow(' — not supported by pdk-ts call yet') : '';
      console.log(`  ${pc.cyan(a.name.padEnd(nameW))}  ${a.type}${support}`);
    }
  }
  console.log();
  console.log(pc.dim(`  pdk-ts call ${pallet} ${call} ${args.map((a) => `<${a.name}>`).join(' ')}`));
  console.log();
}

File: pdk/commands/call.py
"""`pdk call <pallet> <call> [args...]` — generic extrinsic composer.

Every other signing command hardcodes one pallet call. This one doesn't:
it looks up any pallet/call pair in the live runtime's metadata, validates
and coerces the CLI strings against that call's real argument types, and
submits through the same ``submit_call`` / ``receipt_succeeded`` path as
the rest of the toolkit. That inheritance is the point — it picks up the
Assets encoding fix and the event-decode fix rather than inventing a
second signing path that could get either wrong again.

Discovery is built in, because a generic composer is useless if you must
already know the signature to try it::

    pdk call <pallet>          list the pallet's calls and their arg types
    pdk call <pallet> <call>   show that call's expected arguments

Coercion covers a deliberately narrow slice of primitive shapes:
account-like (SS58 or ``//URI``), unsigned integers including
``Compact<...>`` wrapping, bool, and ``Bytes``/``Vec<u8>`` as 0x-hex.
Everything else — structs, enums, ``Option<T>``, tuples, ``Vec<T>`` of
non-bytes, signed integers — is a named, hard error. Never a silent
mis-encode: guessing at an encoding on a signing path is how you sign
something the user didn't ask for.
"""

from __future__ import annotations

import json as jsonlib
import re

import typer
from scalecodec import types
from rich.console import Console
from rich.markup import escape
from rich.table import Table
from substrateinterface import Keypair

from pdk.config import DEFAULT_NODE_URL
from pdk.core.chain import (
    POT_DECIMALS, connect, free_balance, normalise_account_uri, resolve_account, submit_call,
)
from pdk.core.decoder import decode_receipt
from pdk.core.events import receipt_succeeded

console = Console()

_COMPACT = re.compile(r"^Compact<(.+)>$")
# Account-like and byte-string types are matched by name, because what pdk
# has to do with them (SS58/`//URI` resolution, hex validation) doesn't
# follow from their SCALE shape — an AccountId is 32 bytes just like a hash.
_ACCOUNT_TYPES = re.compile(r"^(AccountId32|AccountId|MultiAddress|Address|LookupSource)$", re.IGNORECASE)
_BYTES_TYPES = re.compile(r"^(Bytes|Vec<u8>)$", re.IGNORECASE)
_INT_SCALE_TYPES = (types.U8, types.U16, types.U32, types.U64, types.U128, types.U256)


def unwrap_compact(raw_type: str) -> str:
    """Peel any number of ``Compact<...>`` layers down to the inner type."""
    current = raw_type.strip()
    while True:
        match = _COMPACT.match(current)
        if not match:
            return current
        current = match.group(1).strip()


def classify_type(runtime, raw_type: str) -> str:
    """Classify a metadata type string into a shape we know how to coerce.

    Numbers and bools are identified by asking the chain's own type
    registry what the name resolves to, rather than matching a hand-kept
    list of aliases. This runtime alone declares call arguments as
    `BalanceOf`, `BlockNumber`, `EraIndex`, `Perbill`, `Weight`,
    `AccountIndex`, `BountyIndex`, `RegistrarIndex` and more — all plain
    unsigned integers underneath. A name list would have to guess at each
    one and would silently rot as the runtime adds aliases; the registry
    already knows. Anything that doesn't resolve, or resolves to a
    composite (`Call`, `Vec<T>`, `Option<T>`, structs), stays
    'unsupported' and is refused loudly.
    """
    inner = unwrap_compact(raw_type)
    if _ACCOUNT_TYPES.match(inner):
        return "accountId"
    if _BYTES_TYPES.match(inner):
        return "bytes"
    try:
        resolved = runtime.create_scale_object(inner)
    except Exception:  # noqa: BLE001 — an unknown type name is simply unsupported
        return "unsupported"
    if isinstance(resolved, types.Bool):
        return "bool"
    if isinstance(resolved, _INT_SCALE_TYPES):
        return "integer"
    return "unsupported"


def coerce_arg(kind: str, raw: str, label: str):
    """Coerce one CLI string into what ``compose_call`` expects for ``kind``.

    Raises ``typer.BadParameter`` with the argument's name and declared
    type on bad input, rather than letting a malformed value reach SCALE
    encoding and fail opaquely — or encode something unintended.
    """
    text = raw.strip()
    if kind == "accountId":
        return resolve_account(text)
    if kind == "integer":
        if not text.isdigit():
            raise typer.BadParameter(f"{label}: expected a non-negative integer, got {raw!r}.")
        return int(text)
    if kind == "bool":
        if text.lower() not in ("true", "false"):
            raise typer.BadParameter(f'{label}: expected "true" or "false", got {raw!r}.')
        return text.lower() == "true"
    if kind == "bytes":
        if not re.fullmatch(r"0x[0-9a-fA-F]*", text):
            raise typer.BadParameter(f"{label}: expected 0x-prefixed hex, got {raw!r}.")
        return text
    raise typer.BadParameter(f"{label}: unsupported argument type — pdk call can't encode this yet.")


def _find(names, wanted: str) -> str | None:
    """Case-insensitive name match, mirroring how `pallets`/`storage` resolve."""
    target = wanted.lower()
    return next((n for n in names if n.lower() == target), None)


def _pallet(substrate, name: str):
    return next((p for p in substrate.metadata.pallets if str(p.name) == name), None)


def _call_args(substrate, pallet: str, call: str) -> list[tuple[str, str]]:
    """(name, declared type) for one call, straight off runtime metadata.

    Read from `substrate.metadata`, NOT `get_metadata_call_function`: the
    two disagree. The latter returns the runtime's raw Rust type strings
    (`<T::Lookup as StaticLookup>::Source`, `Compact<T::Balance>`), which
    resolve against nothing; `metadata.pallets` gives the registry-facing
    names (`LookupSource`, `Compact<Balance>`) that classification and
    encoding both need.
    """
    entry = _pallet(substrate, pallet)
    found = next((c for c in (entry.calls or []) if str(c.name) == call), None) if entry else None
    return [(str(a.name), str(a.type)) for a in (getattr(found, "args", None) or [])]


def run(
    pallet: str = typer.Argument(..., help="Pallet name, e.g. Balances."),
    call: str = typer.Argument(None, help="Call name. Omit to list the pallet's calls."),
    args: list[str] = typer.Argument(None, help="Call arguments, in metadata order."),
    sender: str = typer.Option("//Alice", "--from", help="Signing account URI."),
    node: str = typer.Option(DEFAULT_NODE_URL, "--node", help="Portaldot node WS endpoint."),
    dry_run: bool = typer.Option(False, "--dry-run", help="Show the composed call and its fee, submit nothing."),
    json_out: bool = typer.Option(False, "--json", help="Emit machine-readable JSON."),
) -> None:
    """Compose, sign, and submit any pallet call the runtime exposes."""
    try:
        substrate = connect(node)
        substrate.init_runtime()
    except Exception as exc:  # noqa: BLE001
        _die(json_out, f"Cannot reach a Portaldot node at {node}", str(exc),
             "Start a node with pdk up (run the node in WSL on Windows; pdk itself runs natively).")

    pallet_names = [str(p.name) for p in substrate.metadata.pallets]
    resolved_pallet = _find(pallet_names, pallet)
    if resolved_pallet is None:
        _die(json_out, f'Pallet "{pallet}" not found — run pdk pallets to list them.')

    call_names = [str(c.name) for c in (_pallet(substrate, resolved_pallet).calls or [])]

    # Discovery: no call named — list what this pallet offers.
    if not call:
        _render_calls(substrate, resolved_pallet, call_names, json_out)
        return

    resolved_call = _find(call_names, call)
    if resolved_call is None:
        _die(json_out, f'Call "{call}" not found on pallet "{resolved_pallet}" — '
                       f"run pdk call {resolved_pallet} to list its calls.")

    signature = _call_args(substrate, resolved_pallet, resolved_call)
    supplied = list(args or [])

    # Discovery: call named but no args given, and it needs some — show the
    # signature instead of failing on an arg-count error the user can't yet read.
    if not supplied and signature:
        _render_signature(substrate.runtime_config, resolved_pallet, resolved_call, signature, json_out)
        return

    if len(supplied) != len(signature):
        detail = ", ".join(f"{n}: {t}" for n, t in signature) or "no arguments"
        _die(json_out, f"{resolved_pallet}.{resolved_call} expects {len(signature)} argument(s) "
                       f"({detail}), got {len(supplied)}.")

    params = {
        name: coerce_arg(classify_type(substrate.runtime_config, declared), value, f"{name} ({declared})")
        for (name, declared), value in zip(signature, supplied)
    }

    keypair = Keypair.create_from_uri(normalise_account_uri(sender))
    if dry_run:
        _dry_run(substrate, keypair, resolved_pallet, resolved_call, params, json_out)
        return

    try:
        receipt = submit_call(substrate, keypair, resolved_pallet, resolved_call, params)
    except Exception as exc:  # noqa: BLE001
        _die(json_out, f"{resolved_pallet}.{resolved_call} failed: {exc}")

    tx = str(receipt.extrinsic_hash)
    if receipt_succeeded(substrate, receipt):
        if json_out:
            typer.echo(jsonlib.dumps({"success": True, "pallet": resolved_pallet, "call": resolved_call,
                                      "tx": tx, "block": receipt.block_hash}))
        else:
            console.print(f"[green]✓ {escape(resolved_pallet)}.{escape(resolved_call)}[/green]")
            console.print(f"[dim]tx: {tx}[/dim]")
        return

    decoded = decode_receipt(receipt, substrate)
    name = decoded.name if decoded else "unknown"
    if json_out:
        typer.echo(jsonlib.dumps({"success": False, "pallet": resolved_pallet, "call": resolved_call,
                                  "tx": tx, "error": name}))
    else:
        console.print(f"[red]✗ {escape(resolved_pallet)}.{escape(resolved_call)} failed: {escape(name)}[/red]")
        console.print(f"[dim]Diagnose it: pdk debug {tx}[/dim]")
    raise typer.Exit(code=1)


def _dry_run(substrate, keypair, pallet: str, call: str, params: dict, json_out: bool) -> None:
    """Fee-payability preview only. pdk cannot predict arbitrary pallet
    dispatch logic, so this deliberately claims nothing about whether the
    call itself would succeed — only whether the signer can pay for it."""
    try:
        composed = substrate.compose_call(call_module=pallet, call_function=call, call_params=params)
        fee = substrate.get_payment_info(call=composed, keypair=keypair)["partialFee"] / 10**POT_DECIMALS
        balance = free_balance(substrate, keypair.ss58_address)
    except Exception as exc:  # noqa: BLE001
        _die(json_out, f"Could not estimate {pallet}.{call}: {exc}")

    payable = balance >= fee
    if json_out:
        typer.echo(jsonlib.dumps({"dryRun": True, "pallet": pallet, "call": call,
                                  "args": {k: str(v) for k, v in params.items()},
                                  "fee": fee, "senderBalance": balance, "feasible": payable}))
    else:
        console.print(f"[bold]pdk call {escape(pallet)} {escape(call)} --dry-run[/bold]  [dim](not submitted)[/dim]")
        for key, value in params.items():
            console.print(f"  {key:<20} {escape(str(value))}")
        console.print(f"  {'Estimated fee':<20} {fee:.6f} POT")
        console.print(f"  {'Sender balance':<20} {balance:,.4f} POT")
        console.print(f"  {'Prediction':<20} " + ("[green]fee payable[/green]" if payable
                                                  else "[red]would FAIL — cannot cover the fee[/red]"))
        console.print("[dim](fee-payability only — pdk can't predict arbitrary pallet dispatch logic)[/dim]")
    if not payable:
        raise typer.Exit(code=1)


def _render_calls(substrate, pallet: str, call_names: list[str], json_out: bool) -> None:
    rows = [(name, _call_args(substrate, pallet, name)) for name in sorted(call_names)]
    if json_out:
        typer.echo(jsonlib.dumps({"pallet": pallet, "calls": [
            {"name": name, "args": [{"name": n, "type": t} for n, t in sig]} for name, sig in rows
        ]}))
        return
    table = Table(title=f"{pallet} — {len(rows)} calls")
    table.add_column("call", style="green")
    table.add_column("arguments")
    for name, sig in rows:
        table.add_row(escape(name), escape(", ".join(f"{n}: {t}" for n, t in sig)))
    console.print(table)
    console.print(f"[dim]Run one: pdk call {pallet} <call> <args...>[/dim]")


def _render_signature(runtime, pallet: str, call: str, signature: list[tuple[str, str]], json_out: bool) -> None:
    if json_out:
        typer.echo(jsonlib.dumps({"pallet": pallet, "call": call,
                                  "args": [{"name": n, "type": t, "kind": classify_type(runtime, t)} for n, t in signature]}))
        return
    table = Table(title=f"{pallet}.{call}")
    table.add_column("argument", style="cyan")
    table.add_column("type")
    table.add_column("accepts")
    accepts = {
        "accountId": "//URI or SS58 address",
        "integer": "non-negative integer",
        "bool": "true / false",
        "bytes": "0x-prefixed hex",
        "unsupported": "[red]unsupported by pdk call[/red]",
    }
    for name, declared in signature:
        table.add_row(escape(name), escape(declared), accepts[classify_type(runtime, declared)])
    console.print(table)


def _die(json_out: bool, message: str, detail: str = "", hint: str = "") -> None:
    """Emit an error and exit 1 — as JSON under --json so every exit path
    stays parseable, as Rich text otherwise. Always raises."""
    if json_out:
        payload = {"error": message}
        if detail:
            payload["detail"] = detail
        typer.echo(jsonlib.dumps(payload))
    else:
        console.print(f"[red]{escape(message)}[/red]")
        if detail:
            console.print(f"[dim]{escape(detail)}[/dim]")
        if hint:
            console.print(hint)
    raise typer.Exit(code=1)
